### PR TITLE
Change the return type of "extract" and implement "teeWith"

### PR DESCRIPTION
**Pull request by harendra-kumar**
See the commit messages for details. Its better to review each commit individually.

### CODE REVIEW

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 1]
```diff
@@ -1077,22 +1077,17 @@ splitParse (Parser pstep initial extract) (Stream step state) =
...
+                            ParseChunksYield b (ParseChunksInit src s)
             Skip s -> return $ Skip $ ParseChunksStream s buf pst
             Stop   -> do
-                res <- extract pst
-                case res of
-                    Left err -> error err
-                    Right a -> do
-                        let src = Prelude.reverse buf
-                        return $ Skip $ ParseChunksYield a (ParseChunksInitLeftOver src)
+                b <- extract pst
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD.hs:1087]:
```
Stop   -> do
    b <- extract pst
    let src = Prelude.reverse buf
    return $ Skip $
        ParseChunksYield b (ParseChunksInitLeftOver src)
```
The extract at the end seems inconsistent here. I would assume the parsers that go into `splitParse` are terminating and would give a proper result on either `Yield` or `Stop`. Using `extract` here seems inconsistent.  We could save the most recent `Yield` state, which should be extracted, given the stream ends. I don't think that is the behavior right now.

Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD.hs:1087]: `extract` is implemented by the parser, it knows whether there is a consistent partial result available or not, if the parser is implemented correctly extract is guaranteed to  either return a partial result from the last yield or an error. It is doing exactly what you are expecting it to do.

File: src/Streamly/Internal/Data/Parser/Tee.hs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+-- Distributing Parsers
+-------------------------------------------------------------------------------
+
+data StepState s a = StepState s | StepResult a
+
+-- | State of the pair of parsers in a tee composition
+-- Note: strictness annotation is important for fusing the constructors
+{-# ANN type TeeState Fuse #-}
+data TeeState sL sR x a b =
+    TeePair !([x], StepState sL a, [x], [x]) !([x], StepState sR b, [x], [x])
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/Parser/Tee.hs:52]: I think it'd be helpful if you can document what all the arguments here are supposed to be, would be useful while visiting this function in the future.

File: src/Streamly/Internal/Data/Parser/Tee.hs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+-- XXX: With the current "Step" semantics, it is hard to write, and not sure
+-- how useful, an efficient teeWith that returns a correct unused input count.
+--
+-- | @teeWith f p1 p2@ distributes its input to both @p1@ and @p2@ until both
+-- of them succeed or fail and combines their output using @f@. The parser
+-- succeeds if both the parsers succeed.
+--
+-- /Internal/
+--
```
Comment (pranaysashank) [inline src/Streamly/Internal/Data/Parser/Tee.hs:64]: Just a reminded that this function's implementation is complicated, we shouldn't expose this combinator from non-internal modules without tests.

**Comment by harendra-kumar** [comment]
I will merge this and update the comments once when all the PRs are merged.

**Comment by harendra-kumar** [comment]
This has been merged, I have to incorporate some comments from the review, which I will do once all the parser's PRs are merged and then close this.

File: src/Streamly/Internal/Data/Parser.hs [hunk 4]
```diff
@@ -118,33 +127,31 @@ all predicate = Parser step initial (return . Right)
 --
 -- /Internal/
 --
```
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Parser.hs:129]: This has been take care of in the next PR.

File: src/Streamly/Internal/Data/Parser.hs [hunk 5]
```diff
@@ -118,33 +127,31 @@ all predicate = Parser step initial (return . Right)
...
             ++ " elements, got " ++ show i
 
-        f x =
-            if n == i
-            then Right x
-            else Left err
-
 -- | Take at least @n@ input elements, but can collect more.
 --
 -- * Stops - never.
```
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Parser.hs:157]: The fold never terminates, it never returns a Stop result. When the stream ends the parse driver needs to extract the last value. Not sure if the confusion here can be addressed succinctly.

File: src/Streamly/Internal/Data/Stream/StreamD.hs [hunk 6]
```diff
@@ -1077,22 +1077,17 @@ splitParse (Parser pstep initial extract) (Stream step state) =
                             src  = Prelude.reverse src0
                         return $ Skip $ ParseChunksBuf src s buf1 pst1
                     -- XXX Halt 0 common case?
```
[the remark this answers is not in the document]
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Stream/StreamD.hs:1079]: will do.

File: src/Streamly/Internal/Data/Parser/Tee.hs (hunk 2 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Parser/Tee.hs:52]: yes.

**Comment by harendra-kumar** [comment]
Addressed all the comments in #479 .

**Pull request closed without merging**